tools/serve/serve.py

```diff
@@ -204,6 +206,56 @@ def _script_replacement(self, key, value):
         return None
 
 
+class HtmlScriptInjectorHandlerWrapper:
+    def __init__(self, inject="", wrap=None):
+        self.inject = inject
+        self.wrap = wrap
+
+    def __call__(self, request, response):
+        self.wrap(request, response)
+        # If the response content type isn't html, don't modify it.
+        if not isinstance(response.headers, ResponseHeaders) or response.headers.get("Content-Type")[0] != b"text/html":
+            return response
+
+        # Otherwise, inject the script after the document's doctype.
+        data = "".join(item.decode(response.encoding) for item in response.iter_content(read_file=True))
```

Comment:
It has just occured to me that this is going to interact badly with things that use the pipes feature to affect response timing, because that relies on `sleep` calls during each iteration to control when bytes are sent. We already have some features that don't interact well with that, so it's not _essential_ to fix, but we can do better.

Doing a really proper fix is somewhat difficult because I don't think the html5lib tokenizer has an easy way to add new content to the input stream a bit at a time, at least without ending up with very small chunks. That means we can't just iterate over the content, parsing as we go, until we reach the injection point.

The simpler (albeit still partial) is to check if `response.content` is bytes, a string, or has a `read` property, and if it doesn't match any of those, just return the resource as-is. I think that will be unlikely to break in practice because it's uncommon for a resource that does actual DOM API calls to be doing something fancy with an iterator for `content`.

Reply:
I think I've got what you suggested now, I had to include IOBase as FileHandler returns a file object as response.content. Please take a look, thanks.

Reply:
That seems fine; I would have thought that `IOBase` was unnecessary since we have the `read` attribute check (but maybe these days using `IOBase` is better; I was just using the logic from `response.py` in wptserve).